### DISCUSSION TITLE
Why my yarn package not detected in SBOM?

### DISCUSSION THREAD

**Discussion by roah-chae** [aquasecurity/trivy#5396] [Q&A] [answered]
### Question

Hello. 
I am using a docker image to generate SBOM (cyclonedx, spdx-json format).

This is my container environment.
- ubuntu 20.04
- Node.js v16
- package manager: yarn v3.6.4 

When I create a sbom via docker image using yarn v3 package manager, it doesn't detect the node.js package?
The docker images using npm have all detected the package fine.
I have installed lodash, express packages and pino-pretty, pretter as dev dependencies, but only yarn is detected in the SBOM output.

```bash
trivy image --scanners vuln --format cyclonedx --output cdx.json yarn-test:latest
trivy image --format spdx-json --output spdx.json yarn-test:latest
```

Could I be doing something wrong? Or is it hard to detect the yarn package?
This is my github project repository [Link](https://github.com/roah-chae/yarn-berry-sbom-test)

Please help. Thanks.

### Target

SBOM

### Scanner

None

### Output Format

SPDX

### Mode

Standalone

### Operating System

Ubuntu 20.04

### Version

```bash
Version: 0.46.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-10-18 00:20:22.642992927 +0000 UTC
  NextUpdate: 2023-10-18 06:20:22.642992427 +0000 UTC
  DownloadedAt: 2023-10-18 02:57:24.989027 +0000 UTC
```

**Comment by DmitriyLewen** [comment]
Hello @roah-chae 
Thanks for your report.

`image` mode doesn't support `yarn.lock` files.
More information about this [here](https://aquasecurity.github.io/trivy/v0.46/docs/coverage/language/#supported-languages)

Regards, Dmitriy

**Reply by DmitriyLewen** [reply]
You can use [yarn install](https://classic.yarnpkg.com/lang/en/docs/cli/install/) to download all your dependencies into the `node_modules` folder.
Then Trivy can detect your dependencies from `node_modules/path/to/package/package.json` files.

**Reply by roah-chae** [reply]
@DmitriyLewen  Thanks for reply!
According to the [document](https://aquasecurity.github.io/trivy/v0.46/docs/coverage/language/#supported-languages), it seems to be analyzed through package.json as an image, but is it difficult to detect the Yarn package with package.json?
Additionally, Yarn Berry replaces node_modules through the cache folder in the .yarn folder, but is it this difficult to detect?

**Reply by DmitriyLewen** [reply, marked as answer]
`package.json` files use version ranges in `dependencies` field. Therefore, we can't get all dependencies from root `package.json` file. We only scan name and version from each `package.json` files (main file and all files from `node_modules`).
But in your case `yarn` packs packages into zip archives.
Trivy doesn't currently check archives from `.yarn` directory.

@knqyf263 it looks like we need to scan `yarn.lock` files in `image` mode or add support of `.yarn/*.zip` archives to find `package.json` files inside them (we can upgrade our [logic](https://github.com/aquasecurity/trivy/blob/00f2059e5d7bc2ca2e3e8b1562bdfede1ed570e3/pkg/fanal/analyzer/language/nodejs/yarn/yarn.go#L359) for find licenses).
wdyt?

**Reply by roah-chae** [reply]
Thank you.
I hope that the feature will be provided in the future. 🙇
